### ISSUE TITLE
Error when connectiong callback for "drag-data-received"

### ISSUE THREAD

**Issue by lscrd** [StefanSalewski/gintro#73]
When connecting this signal to a callback, the compiler crashes in macros.nim with message:
`Error: expected: ')', but got: ';' [ValueError]`

Here is a simple example to show the issue:
```nim
import gintro/[gobject, gdk, gtk]

proc onDataReceived(window: Label;
                    context: DragContext;
                    x, y: int;
                    seldata: SelectionData;
                    info: int;
                    time: int;
                    data: pointer) = discard

var label = newLabel("")
label.connect("drag-data-received", onDataReceived, pointer(nil))
```

I tried to understand what happened and, finally, looking into the file "gisup3.nim", I found this line:
```
"drag_data_received!Widget!6!(self: Widget; context: gdk.DragContext; x: int; y: int; data: SelectionData;info: int; time: int)!(self: ptr Widget00; context: ptr gdk.DragContext00; x: int32; y: int32;data: ptr SelectionData00; info: uint32; time: uint32)",
```

I decided to add a missing space before the first *info* and before the second *data*. I didn’t expect an improvement, but, surprisingly, there is a change. If there is still a crash, the error message becomes:
`Error: attempt to redefine: 'data'`

So, I renamed *data* to *seldata*. The line became:
```
"drag_data_received!Widget!6!(self: Widget; context: gdk.DragContext; x: int; y: int; seldata: SelectionData; info: int; time: int)!(self: ptr Widget00; context: ptr gdk.DragContext00; x: int32; y: int32; seldata: ptr SelectionData00; info: uint32; time: uint32)",
```

And then, the program compiles. I don’t know if the more complicated code in my program will execute correctly as I have not yet completed the conversion from gtk2 to gintro. But, for sure, there is something wrong with the way the signal "drag-data-received" is handled (and probably some other signals too).

As I don’t a lot of knowledge about the way the connection to callbacks is handled, I can’t say for sure that my changes make sense, especially the addition of these supposed missing spaces. But, I hope they will give a hint about the cause of the issue.

**Comment by StefanSalewski** [comment]
Thanks for reporting this issue.

Unfortunately I have just today shipped version 0.7.4 of gintro with support for latest gtk 3.98.3 and the desired gst message-type support. So I will need a break now, and the macro stuff is really not easy, I have not touched it in the last two years. Will try to investigate it next weekend.

**Comment by lscrd** [comment]
Yes, I’m aware of this new version and, in fact, I have already installed it. There’s no urgency for the issue as I have found a way to solve it. And I have still to convert some other modules.

"gintro" is much more elegant than the low level gtk bindings, but there is a lot of magic in it and when there is a problem, it it not always easy to find the reason. In this case, I had to make sure it should work and to convince myself it was not some mistake from my side.

But, anyway, thanks for your impressive work.

**Comment by StefanSalewski** [comment]
I guess you have already found the cause for the error:

For debugging the connect() macros in gimpl.nim there are two commented our echo statements, activating them gives:

```
proc connect_for_signal_cdecl_drag_data_received1(self: ptr Widget00; context: ptr gdk.DragContext00; x: int32; y: int32;data: ptr SelectionData00; info: uint32; time: uint32; data: pointer) {.cdecl.} =
  let h: pointer = g_object_get_qdata(self, Quark)
  var context1 {.global.}: gdk.DragContext
  if context1.isNil:
    new context1
    context1.ignoreFinalizer = true
  context1.impl = context
  var y: int32;data1 {.global.}: SelectionData
  if y: int32;data1.isNil:
    new y: int32;data1
    y: int32;data1.ignoreFinalizer = true
  y: int32;data1.impl = y: int32;data
  onDataReceived(cast[gtk.Label](h), context1, int(x), y: int32;data1, int(info), int(time), cast[ptr pointer](data)[])

stack trace: (most recent call last)
/home/stefan/.nimble/pkgs/gintro-0.7.4/gintro/gimpl.nim(171, 21) mconnect
/home/stefan/Nim/lib/core/macros.nim(566, 17) parseStmt
/tmp/hhh/t.nim(12, 6) template/generic instantiation of `connect` from here
/home/stefan/Nim/lib/core/macros.nim(566, 17) Error: unhandled exception: /home/stefan/Nim/lib/core/macros.nim(577, 65) Error: expected: ')', but got: ';' [ValueError]
```

So indeed the missing space is the problem,  y: int32;data1 is considered as one entity. Will try to fix that soon.

**Comment by StefanSalewski** [comment]
I hope it is fixed now with version 0.7.6

Indeed there where two errors -- first was a missing space from text reformating, and second was the name conflict of optional parameter of connect macro. You found both errors already and you where even close to fix them!

Note that connect() macro can be also used without the optional parameter, so this compiles too:

```nim
import gintro/[gobject, gdk, gtk]

proc onDataReceived(window: Label;
                    context: DragContext;
                    x, y: int;
                    seldata: SelectionData;
                    info: int;
                    time: int) = discard

var label = newLabel("")
label.connect("drag-data-received", onDataReceived)
```

**Comment by lscrd** [comment]
I have installed the version 0.7.6 and my program compiles (and works). I didn’t know that the extra data parameter was optional. In most cases, I have no use of it, so it will simplify a bit my code by removing this unused parameter.

Thanks a lot for your reactivity.
